README
NAME
    Dist::Zilla::PluginBundle::HARTZELL - My standard dzil config.

VERSION
    version 0.007

SYNOPSIS
       # in dist.ini
       [@HARTZELL]

DESCRIPTION
    This, a Dist::Zilla PluginBundle that builds things the way that I do,
    is a work in progress.

    I left my standard PluginBundle behind when I left my previous job and
    I've discovered that I've fallen behind the state of the art. This is my
    attempt to catch back up. After browsing the bevy of bundles on CPAN I
    decided to model mine on DAGOLDEN's (I like his use of ConfigSlicer and
    various config options). As it stands now it's nearly a copy of his work
    but as my personal preferences assert themselves I expect it to diverge.
    For now I have a lot of github and Meta info stuff to catch up on.

    In its default form it is roughly equivalent to the following dist.ini:

       ; version provider
       [Git::NextVersion]  ; get version from last release tag
       version_regexp = ^release-(.+)$
 
       ; choose files to include
       [Git::GatherDir]         ; everything from git ls-files
       exclude_filename = README.pod   ; skip this generated file
       exclude_filename = META.json    ; skip this generated file
       exclude_filename = cpanfile     ; skip this generated file
 
       [PruneCruft]        ; default stuff to skip
       [ManifestSkip]      ; if -f MANIFEST.SKIP, skip those, too
 
       ; file modifications
       [PkgVersion]        ; add $VERSION = ... to all files
       [PodWeaver]         ; generate Pod
       config_plugin = @HARTZELL
 
       ; generated files
       [ReadmeFromPod]     ; from Pod (runs after PodWeaver)
       [License]           ; boilerplate license
       [ReadmeAnyFromPod]  ; create README.md in repo directory
       type = pod          ; this makes github happy....
       filename = README.pod
       location = root
 
       ; t tests
       [Test::Compile]     ; make sure .pm files all compile
       fake_home = 1       ; fakes $ENV{HOME} just in case
 
       ; xt tests
       [Test::PodSpelling] ; xt/author/pod-spell.t
       [Test::Perl::Critic]; xt/author/critic.t
       [MetaTests]         ; xt/release/meta-yaml.t
       [PodSyntaxTests]    ; xt/release/pod-syntax.t
       [PodCoverageTests]  ; xt/release/pod-coverage.t
       [Test::Portability] ; xt/release/portability.t (of file name)
       [Test::Version]     ; xt/release/test-version.t
 
       ; metadata
       [MinimumPerl]       ; determine minimum perl version
       [AutoPrereqs]       ; find prereqs from code
       skip = ^t::lib
       [MetaNoIndex]       ; sets 'no_index' in META
       directory = t
       directory = xt
       directory = examples
       directory = corpus
       package = DB        ; just in case
 
       [MetaProvides::Package] ; add 'provides' to META files
       meta_noindex = 1        ; respect prior no_index directives
 
       [AutoMetaResources] ; set META resources
       bugtracker.github  = user:hartzell
       repository.github  = user:hartzell
       homepage           = https://metacpan.org/release/%{dist}
 
       [MetaYAML]          ; generate META.yml (v1.4)
       [MetaJSON]          ; generate META.json (v2)
 
       ; build system
       [ExecDir]           ; include 'bin/*' as executables
       [ShareDir]          ; include 'share/' for File::ShareDir
       [Module::Build]     ; create Build.PL
 
       ; copy META.json back to repo dis
       [CopyFilesFromBuild]
       copy = cpanfile
 
       ; manifest (after all generated files)
       [Manifest]          ; create MANIFEST
 
       ; before release
       [Git::Check]        ; ensure all files checked in
       allow_dirty = dist.ini
       allow_dirty = Changes
       allow_dirty = README.pod
 
       [CheckMetaResources]     ; ensure META has 'resources' data
       [CheckPrereqsIndexed]    ; ensure prereqs are on CPAN
       [CheckChangesHasContent] ; ensure Changes has been updated
       [CheckExtraTests]   ; ensure xt/ tests pass
       [TestRelease]       ; ensure t/ tests pass
       [ConfirmRelease]    ; prompt before uploading
 
       ; releaser
       [UploadToCPAN]      ; uploads to CPAN
 
       ; after release
       [Git::Commit / Commit_Dirty_Files] ; commit Changes (as released)
       allow_dirty = dist.ini
       allow_dirty = Changes
       allow_dirty = README.pod
 
       [Git::Tag]          ; tag repo with custom tag
       tag_format = release-%v
 
       ; NextRelease acts *during* pre-release to write $VERSION and
       ; timestamp to Changes and  *after* release to add a new {{$NEXT}}
       ; section, so to act at the right time after release, it must actually
       ; come after Commit_Dirty_Files but before Commit_Changes in the
       ; dist.ini.  It will still act during pre-release as usual
 
       [NextRelease]
 
       [Git::Commit / Commit_Changes] ; commit Changes (for new dev)
       commit_msg = bump Changes
 
       [Git::Push]         ; push repo to remote
       push_to = origin

  Usage
    To use this PluginBundle, just add it to your dist.ini. See the
    "ATTRIBUTES" section for user configurable options.

    This PluginBundle supports ConfigSlicer, so you can pass in options to
    individual plugins directly if necessary.

       [@HARTZELL]
       ExecDir.dir = scripts ; overrides ExecDir's dir option

ATTRIBUTES
  stopwords
    Moose ArrayRef attribute that keeps track of a list of stopwords for
    various spelling tasks. Defaults to an empty list, controlled by
    'stopwords' lines in the dist.ini. There can be multiple stopwords lines
    in the dist.ini file.

    Do this:

      stopwords = one two three
      stopwords = some more words

    to build up your own list of stopwords.

  fake_release
    Moose boolean attribute, when set to true the bundle swaps the
    FakeRelease plugin in place of the UploadToCPAN plugin. Mostly useful
    for testing a dist.ini without risking a real release. Defaults to
    false, controlled by the value of the 'fake_release' dist.ini option.

    Do this:

      fake_release = 1

    to do a fake release.

  no_critic
    Moose boolean attribute, when set to true it disables the Perl::Critic
    tests. Defaults to false, controlled by the value of the 'no_critic'
    dist.ini option.

    Do this:

      no_critic = 1

    to do disable critic testing.

  no_spellcheck
    Moose boolean attribute, when set to true it disables the
    Test::PodSpelling tests. Defaults to false, controlled by the value of
    the 'no_spellcheck' dist.ini option.

    Do this:

      no_spellcheck = 1

    to do disable spell checking.

  is_task
    Moose boolean attribute, when set to true it loads TaskWeaver instead of
    PodWeaver. Defaults to false, controlled by the value of the 'is_task'
    dist.ini option.

    Do this:

      is_task = 1

    to build a Task:: distribution.

  use_autoprereqs
    Moose boolean attribute, when set to true it loads the AutoPrereqs
    plugin. Defaults to true, controlled by the value of the
    'use_autoprereqs' dist.ini option.

    Do this:

      use_autoprereqs = 0

    to do disable AutoPrereqs.

  tag_format
    Moose Str attribute, defines the format used by Git::Tag. Defaults to
    'release-%v', controlled by the value of the 'tag_format' dist.ini
    option.

    Do this:

      tag_format = my-release-%v

    to set it to something other than the default.

  version_regexp
    Moose Str attribute, defines the regexp that Git::NextVersion uses to
    figure out the current version. Defaults to '^release-(.+)$' and is
    controlled by the value of the 'version_regexp' dist.ini option.

    Do this:

      version_regexp = ^my-release-(.+)$

    to set it to something other than the default.

  weaver_config
    Moose Str attribute, controls the name of the PodWeaver configuration.
    Defaults to '@HARTZELL', controlled by the value of the 'weaver_config'
    dist.ini option.

    Do this:

      weaver_config = @DAGOLDEN

    to e.g. use Pod::Weaver::PluginBundle::DAGOLDEN.

  git_remote
    Moose Str attribute, controls the name of the git remote. Defaults to
    'origin', controlled by the value of the 'git_remote' dist.ini option.

    Do this:

      git_remote = 'something_else'

    to use something other than the default.

METHODS
  mvp_multivalue_args
    Returns the list of the dist.ini options that can take multiple values.
    Currently returns qw/stopwords/.

  configure
    Does the heavy lifting, adds the plugins, etc....

COMMON PATTERNS
  nothing much to see here for now....
       [@HARTZELL]
       fakerelease = 1

SEE ALSO
    *   Dist::Zilla

    *   Dist::Zilla::Plugin::PodWeaver

    *   Dist::Zilla::Plugin::TaskWeaver

    *   Dist::Zilla::PluginBundle::ConfigSlicer

AUTHORS
    *   George Hartzell <hartzell@cpan.org>

    *   David Golden <dagolden@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by George Hartzell.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Dist-Zilla-PluginBundle-HARTZELL

0.007     2013-01-24 15:50:47 America/Los_Angeles

  * add Pod::Weaver::PluginBundle::HARTZELL

0.006     2013-01-10 16:13:56 America/Los_Angeles

  * compose both the Easy and Config::Slicer roles in
    a single with statement, better Best Practice.

0.005     2012-12-18 14:11:22 America/Los_Angeles

  * make some prereq recommendations for testing phase
  * touch up abstract

0.004     2012-12-18 10:31:53 America/Los_Angeles

  * make a real cpan release (removed fake_release from dist.ini)

0.003     2012-12-18 10:16:29 America/Los_Angeles

  * no changes, screwed up the last github commit

0.002     2012-12-18 10:14:59 America/Los_Angeles

  * list dependencies in dist.ini, don't 'use' them in
    the bundle.

  * set up stopwords so that the spelling test passes

0.001     2012-12-16 17:33:00 America/Los_Angeles

  * Initial release, basically a fork of DAGOLDEN's bundle with some of
    my personal preferences asserted.


lib/Pod/Weaver/PluginBundle/HARTZELL.pm
use strict;
use warnings;

package Pod::Weaver::PluginBundle::HARTZELL;
{
  $Pod::Weaver::PluginBundle::HARTZELL::VERSION = '0.007';
}

# ABSTRACT: HARTZELL's default Pod::Weaver config


use Pod::Weaver::Config::Assembler;
use Pod::Elemental::Transformer::List 0.101620 ();

#use Pod::Weaver::Section::Support 1.001 ();

sub _exp { Pod::Weaver::Config::Assembler->expand_package( $_[0] ) }

# my $repo_intro = <<'END';
# This is open source software.  The code repository is available for
# public review and contribution under the terms of the license.
# END
#
# my $bugtracker_content = <<'END';
# Please report any bugs or feature requests through the issue tracker
# at {WEB}.
# You will be notified automatically of any progress on your issue.
# END

sub mvp_bundle_config {
    my @plugins;
    push @plugins, (
        [ '@HARTZELL/CorePrep', _exp('@CorePrep'), {} ],
        [ '@HARTZELL/Name',     _exp('Name'),      {} ],
        [ '@HARTZELL/Version',  _exp('Version'),   {} ],

        [ '@HARTZELL/Prelude', _exp('Region'), { region_name => 'prelude' } ],
        [ '@HARTZELL/Synopsis', _exp('Generic'), { header => 'SYNOPSIS' } ],
        [   '@HARTZELL/Description', _exp('Generic'),
            { header => 'DESCRIPTION' }
        ],
        [ '@HARTZELL/Overview', _exp('Generic'), { header => 'OVERVIEW' } ],
    );

    for my $plugin (
        [ 'Attributes', _exp('Collect'), { command => 'attr' } ],
        [ 'Methods',    _exp('Collect'), { command => 'method' } ],
        [ 'Functions',  _exp('Collect'), { command => 'func' } ],
        [ 'Requires',   _exp('Collect'), { command => 'requires' } ],
        )
    {
        $plugin->[2]{header} = uc $plugin->[0];
        push @plugins, $plugin;
    }

    push @plugins, (
        [ '@HARTZELL/Leftovers', _exp('Leftovers'), {} ],
        [   '@HARTZELL/postlude', _exp('Region'),
            { region_name => 'postlude' }
        ],

        #    [ '@HARTZELL/Support',   _exp('Support'),
        #      {
        #        perldoc => 0,
        #        websites => 'none',
        #        bugs => 'metadata',
        #        bugs_content => $bugtracker_content,
        #        repository_link => 'both',
        #        repository_content => $repo_intro
        #      }
        #    ],
        [ '@HARTZELL/Authors', _exp('Authors'), {} ],
        [ '@HARTZELL/Legal',   _exp('Legal'),   {} ],
        [   '@HARTZELL/List', _exp('-Transformer'),
            { 'transformer' => 'List' }
        ],
    );

    return @plugins;
}

1;

__END__

=pod

=head1 NAME

Pod::Weaver::PluginBundle::HARTZELL - HARTZELL's default Pod::Weaver config

=head1 VERSION

version 0.007

=head1 DESCRIPTION

This is a L<Pod::Weaver> PluginBundle.  It is roughly equivalent to
L<Pod::Weaver::PluginBundle::Default> with the addition of

=over 4

=item *

support for collecting "=requires" descriptions into a section titled
"REQUIRES" and

=item *

inclusion of the Transformer plugin.

=back

It is based on Dave Golden's L<Pod::Weaver::PluginBundle::DAGOLDEN>.

=for Pod::Coverage mvp_bundle_config

=head1 USAGE

This PluginBundle is used by default in the
L<Dist::Zilla::PluginBundle::HARTZELL> L<Dist::Zilla> plugin bundle.

=head1 SEE ALSO

=over

=item *

L<Pod::Weaver>

=item *

L<Pod::Elemental::Transformer::List>

=item *

L<Dist::Zilla::Plugin::PodWeaver>

L<Dist::Zilla::PluginBundle::DAGOLDEN>

=back

=head1 AUTHORS

=over 4

=item *

George Hartzell <hartzell@cpan.org>

=item *

David Golden <dagolden@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by George Hartzell.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "File::Find" => 0,
    "File::Temp" => 0,
    "Module::Build" => "0.3601",
    "Test::More" => 0
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "My standard dzil config.",
  "dist_author" => [
    "George Hartzell <hartzell\@cpan.org>",
    "David Golden <dagolden\@cpan.org>"
  ],
  "dist_name" => "Dist-Zilla-PluginBundle-HARTZELL",
  "dist_version" => "0.007",
  "license" => "perl",
  "module_name" => "Dist::Zilla::PluginBundle::HARTZELL",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Dist::Zilla" => "4.3",
    "Dist::Zilla::Plugin::AutoMetaResources" => 0,
    "Dist::Zilla::Plugin::AutoPrereqs" => 0,
    "Dist::Zilla::Plugin::CheckChangesHasContent" => 0,
    "Dist::Zilla::Plugin::CheckExtraTests" => 0,
    "Dist::Zilla::Plugin::CheckMetaResources" => "0.001",
    "Dist::Zilla::Plugin::CheckPrereqsIndexed" => "0.002",
    "Dist::Zilla::Plugin::ConfirmRelease" => 0,
    "Dist::Zilla::Plugin::CopyFilesFromBuild" => 0,
    "Dist::Zilla::Plugin::ExecDir" => 0,
    "Dist::Zilla::Plugin::FakeRelease" => 0,
    "Dist::Zilla::Plugin::Git::Check" => 0,
    "Dist::Zilla::Plugin::Git::Commit" => 0,
    "Dist::Zilla::Plugin::Git::GatherDir" => 0,
    "Dist::Zilla::Plugin::Git::NextVersion" => 0,
    "Dist::Zilla::Plugin::Git::Push" => 0,
    "Dist::Zilla::Plugin::Git::Tag" => 0,
    "Dist::Zilla::Plugin::License" => 0,
    "Dist::Zilla::Plugin::Manifest" => 0,
    "Dist::Zilla::Plugin::ManifestSkip" => 0,
    "Dist::Zilla::Plugin::MetaJSON" => 0,
    "Dist::Zilla::Plugin::MetaNoIndex" => 0,
    "Dist::Zilla::Plugin::MetaProvides::Package" => "1.14",
    "Dist::Zilla::Plugin::MetaTests" => 0,
    "Dist::Zilla::Plugin::MetaYAML" => 0,
    "Dist::Zilla::Plugin::MinimumPerl" => 0,
    "Dist::Zilla::Plugin::ModuleBuild" => 0,
    "Dist::Zilla::Plugin::NextRelease" => 0,
    "Dist::Zilla::Plugin::PkgVersion" => 0,
    "Dist::Zilla::Plugin::PodCoverageTests" => 0,
    "Dist::Zilla::Plugin::PodSyntaxTests" => 0,
    "Dist::Zilla::Plugin::PodWeaver" => 0,
    "Dist::Zilla::Plugin::PruneCruft" => 0,
    "Dist::Zilla::Plugin::ReadmeAnyFromPod" => "0.120051",
    "Dist::Zilla::Plugin::ReadmeFromPod" => 0,
    "Dist::Zilla::Plugin::ShareDir" => 0,
    "Dist::Zilla::Plugin::TaskWeaver" => "0.101620",
    "Dist::Zilla::Plugin::Test::Compile" => 0,
    "Dist::Zilla::Plugin::Test::Perl::Critic" => 0,
    "Dist::Zilla::Plugin::Test::PodSpelling" => "2.001002",
    "Dist::Zilla::Plugin::Test::Portability" => 0,
    "Dist::Zilla::Plugin::Test::Version" => 0,
    "Dist::Zilla::Plugin::TestRelease" => 0,
    "Dist::Zilla::PluginBundle::Filter" => 0,
    "Dist::Zilla::PluginBundle::Git" => "1.121010",
    "Dist::Zilla::Role::PluginBundle::Config::Slicer" => 0,
    "Dist::Zilla::Role::PluginBundle::Easy" => 0,
    "Moose" => "0.99",
    "Pod::Elemental::Transformer::List" => "0.101620",
    "Pod::Weaver::Config::Assembler" => 0,
    "autodie" => "2.00",
    "namespace::autoclean" => "0.09",
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
